### ISSUE TITLE
Multiple "can" for a "with"

### ISSUE THREAD

**Issue by oed** [ucan-wg/spec#123]
Currently it seems like I would need to create multiple `with/can` statements for every action I want to allow for a resource. This means that I would add a lot of duplicate information if I want to delegate a lot of actions for a given resource.[ReCap](https://github.com/spruceid/EIPs/blob/19dae81ecc8354199bbe28f3b6248851f4cf49cf/EIPS/eip-5573.md#recap-details-object-schema) takes a slightly different approach to this were the resource is used as an object key and the value is an array of action strings. 

Maybe I'm missing some trick to represent multiple actions more efficiently?

**Comment by matheus23** [comment]
Actions often imply each other. E.g. for CRUD actions, `can: DELETE` would imply `CREATE`, `READ`, and `UPDATE`, while `can: UPDATE` would imply `can: READ` but not `can: DELETE`.

In CRUD some sets can't be represented easily, e.g. a `CREATE, DELETE` set would need to be two distinct capabilities.
I don't think we've had a need for that in practice though. I guess the set of actions & their implications are just chosen appropriately to avoid having to provide multiple capabilities.

**Comment by Gozala** [comment]
> I don't think we've had a need for that in practice though. I guess the set of actions & their implications are just chosen appropriately to avoid having to provide multiple capabilities.

I think that in by no means implied by the schema. If there is no need to have multiple actions on the same resource than I'd argue `att` schema should be more like:

```ts
"example://example.com/public/photos/": {
   can: "crud/delete",
   nb: {
      "matching": "/(?i)(\W|^)(baloney|darn|drat|fooey|gosh\sdarnit|heck)(\W|$)/"
   }
}
```

However I think ☝️ restrictions would be fairly impractical limitation & supporting multiple actions is inevitable. And in that case resources as keys would indeed be a lot more compact.

> And I would be delighted not to have to deal with `with` property which is [statement in JS](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with) and can't be a variable name.

However [ReCap](https://github.com/spruceid/EIPs/blob/19dae81ecc8354199bbe28f3b6248851f4cf49cf/EIPS/eip-5573.md#recap-details-object-schema) structure is not going to work either, as it does not allow for `nb`. However something along these lines would actually be really nice

```js
{
  "example://example.com/public/photos/": {
      "crud/delete": {}
  },
  "example://example.com/private/84MZ7aqwKn7sNiMGsSbaxsEa6EPnQLoKYbXByxNBrCEr": {
      "wnfs/append": {}
  },
  "example://example.com/public/documents/": {
    "crud/delete": {
      "matching": "/(?i)(\W|^)(baloney|darn|drat|fooey|gosh\sdarnit|heck)(\W|$)/" 
    }
  },
  "mailto:username@example.com": {
    "msg/send": {},
    "msg/receive": {
      "max_count": 5,
      "templates": ["newsletter", "marketing"]
    }
  }
}
```

Which is so much more compact and IMO easier to interpret that original I've taken from the spec:

```js
[
  {
    "with": "example://example.com/public/photos/",
    "can": "crud/delete"
  },
  {
    "with": "example://example.com/private/84MZ7aqwKn7sNiMGsSbaxsEa6EPnQLoKYbXByxNBrCEr",
    "can": "wnfs/append"
  },
  {
    "with": "example://example.com/public/photos/",
    "can": "crud/delete",
    "nb": {
      "matching": "/(?i)(\W|^)(baloney|darn|drat|fooey|gosh\sdarnit|heck)(\W|$)/"
    }
  },
  {
    "with": "mailto:username@example.com",
    "can": "msg/send"
  },
  {
    "with": "mailto:username@example.com",
    "can": "msg/receive",
    "nb": {
      "max_count": 5,
      "templates": ["newsletter", "marketing"]
    }
  }
]
```

Even in the case where no two actions are used with same resource we trim:

1. ` "can"` + `"with"` + `,` and gain `{}` that is we save `11` characters.
2. And if there is `nb`, involved we also save on `"nb":` and no longer pay for `{}` which is we save `18` characters.

I honestly think it would be really nice even if I dread implementing such changes. I also really like that it gets rid of predefined names like `can`, `with` and `nb`, better yet it does not make fields in `nb` feel like they are second class.

**Comment by oed** [comment]
Crossposting for visibility: https://ethereum-magicians.org/t/eip-5573-siwe-recap/10627/21

I think your suggested format could make sense to use in ReCap as well!

**Comment by expede** [comment]
Well, FWIW, I'm totally happy to make it look like the above. We had a previous version that looked roughly like that, but got complaints from JS/TS folks. We've migrated the spec from another other repo, so having a hard time finding anything that far back.

What I like about this approach is that it's very straightforward.

The only thing that I don't love about it is that all of those extra `{}`s are ugly in the simple case.

```json
{
  "example://example.com/public/photos/": {
      "crud/delete": {}
  },
  "example://example.com/private/84MZ7aqwKn7sNiMGsSbaxsEa6EPnQLoKYbXByxNBrCEr": {
      "wnfs/append": {}
  },
  "example://example.com/public/documents/": {
      "crud/delete": {}
  },
  "mailto:username@example.com": {
      "msg/send": {}
  }
}
```

Also note that JSON allows for duplicate fields, but that this behaviour is inconsistent. From RFC 8259 ("The JavaScript Object Notation (JSON) Data Interchange Format"):

> An object whose names are all unique is interoperable in the sense
   that all software implementations receiving that object will agree on
   the name-value mappings.  When the names within an object are not
   unique, the behavior of software that receives such an object is
   unpredictable.  Many implementations report the last name/value pair
   only.  Other implementations report an error or fail to parse the object, 
   and some implementations report all of the name/value pairs, 
   including duplicates.
>
>   JSON parsing libraries have been observed to differ as to whether or
   not they make the ordering of object members visible to calling
   software.  Implementations whose behavior does not depend on member
   ordering will be interoperable in the sense that they will not be
   affected by these differences.

It's not clear from looking at the raw certificate what the intended behaviour is:

```json
{
  "mailto:username@example.com": {
      "msg/send": { "to": "someone@email.com" },
      "msg/receive": {}
  },
  "mailto:username@example.com": {
      "msg/send": {}
  }
}
```

Arguably this should be equivalent to:

```json
{
  "mailto:username@example.com": {
      "msg/send": {},
      "msg/receive": {}
  }
}
```

In any case, this is a difficult case for a lot of parsers.

**Comment by expede** [comment]
> And I would be delighted not to have to deal with with property which is [statement in JS](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with) and can't be a variable name.

Oooh when destructuring gotcha. Yeah that's annoying to have to bind it to something like `const {with: with_, can}`

**Comment by expede** [comment]
> In any case, this is a difficult case for a lot of parsers.

Arguably this is the worst of both worlds, but bear with me:

```json
[
  {
    "cap": {
      "mailto:username@example.com": {
        "msg/send": { "to": "someone@email.com" },
        "msg/receive": {}
    }
  },
  {
    "cap": {
      "mailto:username@example.com": {
        "msg/send": {}
    }
  }
}
```

🤔 this all feels very familiar somehow. I think I may be retracing steps from the last time this came up circa 2020(?) (which would have been quite early in the UCAN days).

**Comment by Gozala** [comment]
> To make it read nicely, we could also make life harder for implementers with something like:
> 
> ```json
> {
>   "mailto:alice@example.com": "msg/send",
>   "mailto:bob@example.com": ["msg/receive", "msg/send"],
>   "mailto:carol@example.com": {"msg/send": {"to": "alice@example.com"}}
> }
> ```
> 
> ...which I'm not a huge fan of TBH

I would much rather keep it simple and live with extra `{}`. ☝️ not only would make parsers bit annoying to write but also for type inference to make a good sense of, which I'm bit sensitive to given how much of our tooling depends on it.

**Comment by expede** [comment]
Disclaimer: I'm still not a fan of this form. Just trying to understand the core problem!

> type inference to make a good sense of, which I'm bit sensitive to given how much of our tooling depends on it.

Wouldn't you model that with a variant, and then normalize it to a consistent representation in the IR?

**Comment by expede** [comment]
> Arguably this is the worst of both worlds, but clean for parsers. Just bear with me during this thought experiment:

It's been all of 25 minutes, but this version is actually kind of growing on me if it eliminates the JSON parser problems. It's a handful of extra characters, but way less ambiguity.

**Comment by expede** [comment]
Ah, I think I've fully walked myself back through to `with/can`. I think the conversation back them went something like this:

This is a bit ugly in the simple case:

```json
[
  {
    "mailto:username@example.com": {
      "msg/send": {},
      "msg/receive": {}
    }
  },
  {
    "mailto:username@example.com": {
      "msg/send": {}
    }
  }
]
```

So okay, let's get rid of those dangling `{}`s that are unneeded the vast majority of the time by un-nesting them:

```
[
  {
    "mailto:username@example.com": [
      {"can": "msg/send", "to": "someone@email.com" },
      {"can": "msg/receive"}
    ]
  },
  {
    "mailto:username@example.com": [
      {can: "msg/send"}
    ]
  }
]
```

Which is two very small steps away from `with/can`

**Comment by oed** [comment]
> It's not clear from looking at the raw certificate what the intended behaviour is:

Simplest thing is probably to say that all UCANs with duplicate resource keys are invalid.

**Comment by matheus23** [comment]
I think Brooke has hit the nail on the head with the above comment.
My blink reaction to the proposal was "yeah that looks amazing to read!", but if you look closer, it turns out you can not represent the same info as with a list of records, unless you allow multiple keys in a record to have the same name.

This is me trying to prepend Brooke's list of examples from above:
```json
{
    "mailto:username@example.com": {
        "msg/send": { "to": "someone@email.com" },
        "msg/receive": {}
    },
    "mailto:username@example.com": {
        "msg/send": { "to": "joe@email.com" }
    }
}
```
This cannot be represented if you disallow duplicate keys and want to avoid making `to` more complicated.

**Comment by Gozala** [comment]
We could alternatively say that

```js
{
  “mailto:username@example.com": {
    "msg/send": {
      “to": "someone@email.com”
     },
   }
}
```

is sugar for 

```js
{
  “mailto:username@example.com": {
    "msg/send": {
      “to": "someone@email.com”
     },
   }
}
```

**Comment by Gozala** [comment]
I think perhaps more serious problem is that today you can just stick capability into an array don’t worry about it.

Proposed form however would require merging capabilities and there is also canonicalization problem (although it’s not any worse than today)

**Comment by expede** [comment]
> I think perhaps more serious problem is that today you can just stick capability into an array don’t worry about it.

Yeah, exactly. This is an advantage of the current model: it is consistent even if it's less DRY.

> Proposed form however would require merging capabilities and there is also canonicalization problem (although it’s not any worse than today)

We do have the option of specifying that duplicate resource keys are not allowed. It ends up being one more thing that implementers have to be careful about, which they don't have to today — i.e. with this change, there would be more ways to accidentally make broken UCANs.

But also if someone had multiple `exp` fields, we would call that malformed. Intuitively, it seems more likely to make this mistake in the capabilities than expiry, though.

> but then however if we switch empty {} with empty [] it’s all going to work out wouldn’t it ?

Works for me 👍 It handles the common case of no extended fields nicely.

> is sugar for

Arguably this is also sugar:

```json
"mailto:username@example.com": [
  {"can": "msg/send", "to": "someone@email.com"},
  "msg/receive"
]
```

Which is valid, but makes it harder to type and reason about. In general, my vote will be with consistency

**Comment by expede** [comment]
Idea from anther thread, but that we have the groundwork for as of 0.9:

```
{
  "example.resource.1": {
    "crud/read": {
      "prf": ["QmABCDEF"]
    },
    "crud/update": {
      "prf": ["Qm12345"]
    }
  }
}
```

This means that you always have a 3-level deep `nb`, which includes the proofs. The downside is

```js
{
  "example.com/foo": {
    "crud/read": {
      "prf": ["QmREVOKED"] // NOTE
    },
    "crud/update": {
      "prf": ["Qm12345"]
    }
  },
  "example.com": {
    "crud/create": {
      "prf": ["QmReadAndCreate"] // Has read and create access
    }
  },
}
```

I have all of the relevant proofs for `"example.com/foo" -> "crud/read"` in this UCAN, but can't use it. We also repeat the token `prf` a lot.

**Comment by expede** [comment]
Okay, I have sign off from the core UCAN crew 👍 Let's try some of these out and see how it feels

**Comment by expede** [comment]
I think this is too far, but to get people's feeling:

```js
{
  "https://example.com": 
    "/foo": {
      "crud/read": {},
      "crud/update": {}
    },
    "/": {
      "crud/create": {
        "mime": ["application/json", "text/html"]
      }
    }
  }
}
```

**Comment by Gozala** [comment]
> I think this is too far, but to get people's feeling:


I have few concerns:

1. I worry it would just make canonicalization bit more error prone. 
2. TS inference is going to either suffer of types will end up more complex & I’d prefer not to complicate them further.
3. Biggest appeal for me was improved readability, I think this is not as readable 

Not saying we should not, just something worth considering.

I also do have to say that splitting abilities did cross my mind too, but then above arguments was a reason I discarded it

**Comment by expede** [comment]
Closed by #132
